### ISSUE TITLE
Inconsistent Specs for the Same Package

### ISSUE THREAD

**Issue by citibeth** [LLNL/spack#2229]
@tgamblin @alalazo @adamjstewart 

I've had some problems with the same package being installed twice under different hashcodes, even though NOTHING relevant changed.  Certain packages seem to be "hot spots" for this behavior; for example, `py-numpy` in my case.  See:

```
$ ls -ltd py-numpy-*
drwxr-xr-x. 5 rpfische rpfische 4096 Nov  2 20:48 py-numpy-1.11.1-4x5i7jdcohqmmygrihehgstdrwhrh5xy
drwxr-xr-x. 5 rpfische rpfische 4096 Oct 12 14:59 py-numpy-1.11.1-gvcr5ncrxern5m7g67q5r5d3p3odbkq5
drwxr-xr-x. 5 rpfische rpfische 4096 Oct  6 13:31 py-numpy-1.11.1-5fjabod6muxily63zp3qjdczj6v3yowg
```
The last change to my Spack was Oct 6, so why does it keep re-installing?  Let's look at the `.spack/spec.yaml` for each of them:

[spec-4x5i7jdcohqmmygrihehgstdrwhrh5xy.yaml.txt](https://github.com/LLNL/spack/files/568735/spec-4x5i7jdcohqmmygrihehgstdrwhrh5xy.yaml.txt)
[spec-gvcr5ncrxern5m7g67q5r5d3p3odbkq5.yaml.txt](https://github.com/LLNL/spack/files/568734/spec-gvcr5ncrxern5m7g67q5r5d3p3odbkq5.yaml.txt)
[spec-5fjabod6muxily63zp3qjdczj6v3yowg.yaml.txt](https://github.com/LLNL/spack/files/568736/spec-5fjabod6muxily63zp3qjdczj6v3yowg.yaml.txt)

Let's compare the most recent two (again, no relevant changes to my Spack in that time).  First of all, a quick look at `spec-gvc...` looks wrong, what is this `&id001` stuff?
```
spec:
- py-numpy:
    version: 1.11.1
    compiler:
      name: gcc
      version: 4.9.3
    namespace: builtin
    parameters:
      blas: true
      lapack: true
      cflags: &id001 []
      cppflags: &id002 []
      cxxflags: &id003 []
      fflags: &id004 []
      ldflags: &id005 []
      ldlibs: &id006 []
```

Second... the dependencies listed in these files are different, even though `py-numpy/packages.yaml` never changed!  For example, `spec-gv..` has `py-nose`, but `spec-5f..` does not.

In my work, from time to time I install the same package from different top-level `spack install` commands.  For example, both `spack setup icebin` and `spack install env-modele` ( https://github.com/citibeth/spack/blob/efischer/develop/var/spack/repos/builtin/packages/env-modele/package.py ) will result in `py-numpy` being installed.  But Spack is apparently installing them under different hash codes (this would be easier to prove directly with #2224).

My best guess about what's going on here... when I say `spack spec`, I get a tree in which every package is listed *exactly once*.  We all know that dependencies form a DAG, not a tree --- where a single dependency can appear along multiple paths from the root.  Spack converts the DAG to tree in an arbitrary way.  The result is that the set of dependencies for the same package will look different in the tree, depending on whether some of its dependencies were "stolen" by other packages elsewhere in the same tree.  Thus, the apparent dependencies of a package will differ depending on the top-level package in the DAG.

It had been my implicit assumption that Spack operated on a DAG, and was just showing it as a tree as a matter of practical convenience for the user when displying `spack spec`.  But... it appears that this tree, no the full DAG, is *actually* what's used in `spec.yaml` and in computing hashes.  That is *clearly* buggy behavior, and would explain a lot of the "mysterious" re-installs we've been experiencing.

This bug should be fixed; relying on already-installed packages, as oft-talked about, will likely not fix this bug.  If we have #2224, then I believe I could prepare a pair of `spack spec` statements that produce a different hash for the same package, thus nailing down an easy test case.  But even without #2224, I'm fairly confident I've figured out the root cause of this mysterious behavior.

[For the record... I do all my install commands with "clean" command lines, i.e. no `^xxx` stuff.  See enclosed `packages.yaml` that controls the versions and variants of things that get installed: 
[packages.yaml.txt](https://github.com/LLNL/spack/files/568753/packages.yaml.txt) ]

**Comment by citibeth** [comment]
I added this to the Project as "nice to have" and not "required."  It is annoying and would be REALLY NICE to get fixed.  But Spack is still usable even with this bug.

**Comment by alalazo** [comment]
Easy answers first :smile: 

> Let's compare the most recent two (again, no relevant changes to my Spack in that time). First of all, a quick look at spec-gvc... looks wrong, what is this &id001 stuff?

Those are `yaml` anchors : they give an `id` to an object (in this case an empty list) and re-use it afterwards with `*` keyword. Note that it means that in the object that was dumped *the same python object* was used for both entries. An example is worth a thousand words here:
```python
>>> import yaml
>>> a = {'a': 1}
>>> b = {1: a, 2: a}
>>> print(yaml.dump(b))
1: &id001 {a: 1}
2: *id001
```

**Comment by alalazo** [comment]
> It had been my implicit assumption that Spack operated on a DAG, and was just showing it as a tree as a matter of practical convenience for the user when displying spack spec. But... it appears that this tree, no the full DAG, is actually what's used in spec.yaml and in computing hashes. That is clearly buggy behavior, and would explain a lot of the "mysterious" re-installs we've been experiencing.

What makes you think `Spack` operates on a tree and not on the full DAG? 

As far as I can see in `spec.yaml` you have all the information to fully reconstruct the DAG (including dependencies for every node, and type of dependencies). Also, don't know if you knew it, but there's another command similar to `spec`:
```console
$ spack graph hdf5
o  hdf5
|\
o |  zlib
 /
o  openmpi
o  hwloc
o  libpciaccess
```
for complex packages it supports `dot` to output graphs.

**Comment by citibeth** [comment]
> What makes you think Spack operates on a tree and not on the full DAG?
>

It's my best hypothesis to explain why `py-nose` is in one `spec.yaml` and
not the other.  The hypothesis could be wrong, of course, but the
mysterious `spec.yaml` files show that something is going wrong.

>
>

**Comment by alalazo** [comment]
Can you post a `tgz` (under `txt` disguise) of the whole `.spack` folder for say, the first installation? There are other inconsistencies the `spec.yaml`: for instance the dependency types are wrong for `py-setuptools` or `py-nose`. @tgamblin Is this something known?

**Comment by citibeth** [comment]
OK, here it is.
[spack.tar.gz.txt](https://github.com/LLNL/spack/files/569022/spack.tar.gz.txt)


BTW... I just renamed `env-modele` to `env-modele-ks` in my repo.  Last night I did `spack install env-modele`, so there shouldn't be much change.  Right?  Well... it's installing a whole slew of stuff, including a new `netcdf`.  I've included that in the tarfile as well, plus a log of what it's installing (so far).

```
$ spack install env-modele-ks
==> Installing env-modele-ks
==> Installing py-udunits
==> python is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/python-3.5.2-fsiwqdnbv37srf4h67uuhxrffixlkwbn
==> py-six is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/py-six-1.10.0-xfewzmfjc3cti25b4deuzb5lznazixmm
==> udunits2 is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/udunits2-2.2.20-quvwfaeuszeojox2p4a3cewxjfzvynuv
==> Installing py-netcdf
==> python is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/python-3.5.2-fsiwqdnbv37srf4h67uuhxrffixlkwbn
==> py-cython is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/py-cython-0.23.5-jf4xoeoxsklr2ld3aqkvfjashkmrlhyj
==> Installing netcdf
==> m4 is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/m4-1.4.17-2remfszqig3xw6d4nqiaenmiqxu2wern
==> curl is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/curl-7.50.3-glayfkqesbk3bfg6imbocknepyoyat2v
==> zlib is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/zlib-1.2.8-ycgkingohwok4tyxiexwhr4o364w2cwn
==> hdf5 is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/hdf5-1.10.0-patch1-skdpe3u7qftoiululwebruccm5me6erx
==> Trying to fetch from file:///home2/rpfische/spack3/var/spack/cache/netcdf/netcdf-4.4.1.tar.gz
######################################################################## 100.0%
==> Staging archive: /home2/rpfische/spack3/var/spack/stage/netcdf-4.4.1-ennixbzq7myx5catlskmfbgjq24y6svz/netcdf-4.4.1.tar.gz
==> Created stage in /home2/rpfische/spack3/var/spack/stage/netcdf-4.4.1-ennixbzq7myx5catlskmfbgjq24y6svz
==> No patches needed for netcdf
==> Building netcdf
==> Successfully installed netcdf
  Fetch: 0.12s.  Build: 53.03s.  Total: 53.14s.
[+] /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/netcdf-4.4.1-ennixbzq7myx5catlskmfbgjq24y6svz
==> Installing binutils
==> m4 is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/m4-1.4.17-2remfszqig3xw6d4nqiaenmiqxu2wern
==> Installing bison
==> m4 is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/m4-1.4.17-2remfszqig3xw6d4nqiaenmiqxu2wern
==> Trying to fetch from file:///home2/rpfische/spack3/var/spack/cache/bison/bison-3.0.4.tar.gz
######################################################################## 100.0%
==> Staging archive: /home2/rpfische/spack3/var/spack/stage/bison-3.0.4-idfvnruiomrwnvb6cjlew6h64nzcdakl/bison-3.0.4.tar.gz
==> Created stage in /home2/rpfische/spack3/var/spack/stage/bison-3.0.4-idfvnruiomrwnvb6cjlew6h64nzcdakl
==> No patches needed for bison
==> Building bison
==> Successfully installed bison
  Fetch: 0.13s.  Build: 1m 31.69s.  Total: 1m 31.82s.
[+] /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/bison-3.0.4-idfvnruiomrwnvb6cjlew6h64nzcdakl
==> Installing flex
==> m4 is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/m4-1.4.17-2remfszqig3xw6d4nqiaenmiqxu2wern
==> bison is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/bison-3.0.4-idfvnruiomrwnvb6cjlew6h64nzcdakl
==> Trying to fetch from file:///home2/rpfische/spack3/var/spack/cache/flex/flex-2.6.0.tar.gz
######################################################################## 100.0%
==> Staging archive: /home2/rpfische/spack3/var/spack/stage/flex-2.6.0-afqykeqxajgrqfnlkpzdoqwcamywr4j7/flex-2.6.0.tar.gz
==> Created stage in /home2/rpfische/spack3/var/spack/stage/flex-2.6.0-afqykeqxajgrqfnlkpzdoqwcamywr4j7
==> No patches needed for flex
==> Building flex
==> Successfully installed flex
  Fetch: 0.09s.  Build: 23.23s.  Total: 23.31s.
[+] /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/flex-2.6.0-afqykeqxajgrqfnlkpzdoqwcamywr4j7
==> Trying to fetch from file:///home2/rpfische/spack3/var/spack/cache/binutils/binutils-2.27.tar.bz2
######################################################################## 100.0%
==> Staging archive: /home2/rpfische/spack3/var/spack/stage/binutils-2.27-swhurdvzsiprppgckwmtq6qcmis7na5y/binutils-2.27.tar.bz2
==> Created stage in /home2/rpfische/spack3/var/spack/stage/binutils-2.27-swhurdvzsiprppgckwmtq6qcmis7na5y
==> Applied patch cr16.patch
==> Building binutils
==> Successfully installed binutils
  Fetch: 0.26s.  Build: 5m 40.77s.  Total: 5m 41.03s.
[+] /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/binutils-2.27-swhurdvzsiprppgckwmtq6qcmis7na5y
==> py-numpy is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/py-numpy-1.11.1-5fjabod6muxily63zp3qjdczj6v3yowg
==> Trying to fetch from file:///home2/rpfische/spack3/var/spack/cache/py-netcdf/py-netcdf-1.2.3.1.tar.gz
######################################################################## 100.0%
==> Staging archive: /home2/rpfische/spack3/var/spack/stage/py-netcdf-1.2.3.1-onhfzeqhd6ogrcxqocwhvdm4oc7zu6yv/v1.2.3.1rel
==> Created stage in /home2/rpfische/spack3/var/spack/stage/py-netcdf-1.2.3.1-onhfzeqhd6ogrcxqocwhvdm4oc7zu6yv
==> No patches needed for py-netcdf
==> Building py-netcdf
==> Successfully installed py-netcdf
  Fetch: 0.24s.  Build: 1m 37.98s.  Total: 1m 38.23s.
[+] /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/py-netcdf-1.2.3.1-onhfzeqhd6ogrcxqocwhvdm4oc7zu6yv
==> Trying to fetch from file:///home2/rpfische/spack3/var/spack/cache/py-udunits/py-udunits-1.1.3.tar.gz
######################################################################## 100.0%
==> Staging archive: /home2/rpfische/spack3/var/spack/stage/py-udunits-1.1.3-r6ddasvrjhxi7azqbklhkj2v6tucg6jp/v1.1.3.tar.gz
==> Created stage in /home2/rpfische/spack3/var/spack/stage/py-udunits-1.1.3-r6ddasvrjhxi7azqbklhkj2v6tucg6jp
==> No patches needed for py-udunits
==> Building py-udunits
==> Successfully installed py-udunits
  Fetch: 0.08s.  Build: 4.09s.  Total: 4.17s.
[+] /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/py-udunits-1.1.3-r6ddasvrjhxi7azqbklhkj2v6tucg6jp
==> Installing modele-utils
==> m4 is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/m4-1.4.17-2remfszqig3xw6d4nqiaenmiqxu2wern
==> Installing cmake
==> bzip2 is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/bzip2-1.0.6-r5epeoxchzk3bzfr7osg7tant4h6nxek
==> expat is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/expat-2.2.0-gs5rqbzqycdab23tvjtfl225ollzu2gl
==> ncurses is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/ncurses-6.0-qtih3zt7fn7hg3sfkeg53ta3u5nhkzp5
==> zlib is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/zlib-1.2.8-ycgkingohwok4tyxiexwhr4o364w2cwn
==> openssl is externally installed in /false/path
==> Installing libarchive
==> lzma is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/lzma-4.32.7-x2zfefshzfic3hzo6bp7vf73ckqdyi4g
==> xz is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/xz-5.2.2-xrfipwm7kpcbqcdp2bdreavtoaaerczb
==> bzip2 is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/bzip2-1.0.6-r5epeoxchzk3bzfr7osg7tant4h6nxek
==> zlib is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/zlib-1.2.8-ycgkingohwok4tyxiexwhr4o364w2cwn
==> libxml2 is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/libxml2-2.9.4-wi2nujzykapy7hesipbcrmqym44vqsdz
==> expat is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/expat-2.2.0-gs5rqbzqycdab23tvjtfl225ollzu2gl
==> lz4 is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/lz4-131-kwt7io2qxeriqn3ekg5fim5odtb37z52
==> openssl is externally installed in /false/path
==> Installing nettle
==> Installing gmp
==> m4 is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/m4-1.4.17-2remfszqig3xw6d4nqiaenmiqxu2wern
==> Trying to fetch from file:///home2/rpfische/spack3/var/spack/cache/gmp/gmp-6.1.1.tar.bz2
######################################################################## 100.0%
==> Staging archive: /home2/rpfische/spack3/var/spack/stage/gmp-6.1.1-v5mkm27wnpwjv2woutpgvglx43j4ogfx/gmp-6.1.1.tar.bz2
==> Created stage in /home2/rpfische/spack3/var/spack/stage/gmp-6.1.1-v5mkm27wnpwjv2woutpgvglx43j4ogfx
==> No patches needed for gmp
==> Building gmp
==> Successfully installed gmp
  Fetch: 3.60s.  Build: 2m 53.29s.  Total: 2m 56.88s.
[+] /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/gmp-6.1.1-v5mkm27wnpwjv2woutpgvglx43j4ogfx
==> Trying to fetch from file:///home2/rpfische/spack3/var/spack/cache/nettle/nettle-3.2.tar.gz
######################################################################## 100.0%
==> Staging archive: /home2/rpfische/spack3/var/spack/stage/nettle-3.2-4bbxbcse3j4dlldrksi2lgangdjg33y4/nettle-3.2.tar.gz
==> Created stage in /home2/rpfische/spack3/var/spack/stage/nettle-3.2-4bbxbcse3j4dlldrksi2lgangdjg33y4
==> No patches needed for nettle
==> Building nettle
==> Successfully installed nettle
  Fetch: 0.10s.  Build: 37.20s.  Total: 37.30s.
[+] /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/nettle-3.2-4bbxbcse3j4dlldrksi2lgangdjg33y4
==> lzo is already installed in /home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/lzo-2.09-dmtdcddklrgsg6qxu43hfl2pz3jzsler
==> Trying to fetch from file:///home2/rpfische/spack3/var/spack/cache/libarchive/libarchive-3.2.1.tar.gz
######################################################################## 100.0%
==> Staging archive: /home2/rpfische/spack3/var/spack/stage/libarchive-3.2.1-hntuo4precse77x4bjtao2472x2si4lf/libarchive-3.2.1.tar.gz
==> Created stage in /home2/rpfische/spack3/var/spack/stage/libarchive-3.2.1-hntuo4precse77x4bjtao2472x2si4lf
==> No patches needed for libarchive
==> Building libarchive
```

**Comment by alalazo** [comment]
@citibeth Summary of facts I noticed:

1. All `packages.py` **are equal**

  ```console
  $ diff -r py-numpy-1.11.1-5fjabod6muxily63zp3qjdczj6v3yowg/.spack/repos/builtin/packages py-numpy-1.11.1-gvcr5ncrxern5m7g67q5r5d3p3odbkq5/.spack/repos/builtin/packages 
  $ diff -r py-numpy-1.11.1-5fjabod6muxily63zp3qjdczj6v3yowg/.spack/repos/builtin/packages py-numpy-1.11.1-4x5i7jdcohqmmygrihehgstdrwhrh5xy/.spack/repos/builtin/packages
  ```
  so the problem is not there.

1. The installation that is apparently "right" for `py-numpy` is `5fjabod6muxily63zp3qjdczj6v3yowg` : build dependencies shouldn't be listed in `spec.yaml` as they don't take part in the hash computation. Can you confirm @mathstuf ?

1. For some reason the other two installations differ in the fact that either `py-setuptools` or `py-nose` are considered to be of `type=('build', 'link')` instead of just `type=('build,)`. This is strange to me, as I can't think of a way in which the edge type can be changed **without** changing `py-numpy/package.py`.

1. There are always two `specs` sharing the same `cflags`, etc. objects and that seems also strange. If those flags are meant to be per-package it shouldn't happen. If they are meant to be global, then why a lot of packages have copies of those?

1. I couldn't reproduce the issue, and as far as I can tell the only things that generally I do differently is that I don't use `packages.yaml` to help the concretizer.

So: I think you are right in saying that something is obviously wrong, but if I could bet I would say that it is in the interaction between `packages.yaml` and the concretizer. I am pretty sure the internal representation is that of a DAG.

**Comment by citibeth** [comment]
>
> I couldn't reproduce the issue, and as far as I can tell the only things
> that generally I do differently is that I don't use packages.yaml to help
> the concretizer.


That sounds like a good clue.  Why don't we focus on getting a minimal case
where things go wrong?  If we can get #2224 implemented, then I can play
around with a clean Spack, try different things in `packages.yaml` and see
if I can reproduce this bug (meaning... get different hashes coming out of
`spack spec` for the same package).

But first step is to implement #2224.



On Thu, Nov 3, 2016 at 9:47 AM, Massimiliano Culpo <notifications@github.com

**Comment by adamjstewart** [comment]
I've also ran into this problem as well. I wish I could provide steps to reproduce it. Like @citibeth, my assumption was that Spack was using a tree instead of DAG internally and that was causing the problem. But if that's not the case, maybe it has something to do with the deptypes? If package A depends on B and C, and B has a 'build' deptype on D while C has a 'build', 'link' deptype on D, are we positive that the DAG shows that B depends on D but C doesn't? Maybe we are accidentally taking the summation of all deptypes when considering a package. Once we figure out what is really going on here we definitely need to add a unit test to prevent it from happening again.

By the way, is there a quick and easy way to add fake packages to Spack for testing purposes. I just want to create packages A-D and see what happens when I build A, B, and C individually. Theoretically, if I build A first, B and C would already be installed.

**Comment by citibeth** [comment]
I believe there are testing-oriented packages in the test repo.

I also have a "bundle" package (env-modele-ks) in efischer/develop on my
fork (github/citibeth/spack).  It's pretty fast to install.

**Comment by adamjstewart** [comment]
How do I create and use testing (mock?) packages? Just another thing that should be documented in the Developer Guide... _sigh_

**Comment by alalazo** [comment]
Have you tried to use the `fake` option of `install`?

**Comment by citibeth** [comment]
No.... that sounds like a good idea, and less work than adding a new
feature.  I'll see what I can do with it.

On Thu, Nov 3, 2016 at 1:47 PM, Massimiliano Culpo <notifications@github.com

**Comment by adamjstewart** [comment]
> Have you tried to use the `fake` option of `install`?

Wow, how long has that been there?

**Comment by tgamblin** [comment]
>> Have you tried to use the fake option of install?

> Wow, how long has that been there?

A while :)  It's for debugging.

**Comment by tgamblin** [comment]
Ok so it looks *to me* that the issue here is the YAML anchors.  They will adversely affect the hash in ways we don't want.  Ugh.  I looked at this previously and I thought I had gotten rid of them by duplicating certain objects in specs.  But apparently not.  I dug a bit deeper and it looks like we can [force YAML to ignore them](http://pyyaml.org/ticket/91).  So I *think* that is the fix.

Some clarifications:

1. Spack operates on a DAG, not a tree.  `spec.yaml` encodes the DAG.
1. There are actually options for how to draw that tree.  Look at [`spec.traverse()`](https://github.com/LLNL/spack/blob/develop/lib/spack/spack/spec.py#L844).  The options are not on the `spec` command, but you can choose to have the traversal "cover" unique nodes, edges, or paths.  The default is `nodes` because it draws each node only once, but the `edges` option might be what you want.  There is also spack graph but it's hard to read for large things.
1. We currently intentionally do not store build dependencies in the `spec.yaml` files.  This is a little awkward implementation-wise, but it is intended to partially address the concerns in #1325 while we wait for a better concretizer.  Not hashing build dependencies basically makes it so that hashes don't change every time something in the tree starts depending on `cmake` or `automake`, and likewise Spack won't rebuild a packages that transitively depend on things like that.
1. The DAG hash is computed by generating a canonical version of `spec.yaml` (basically spec.yaml without the hash fields), running `SHA-1` on that, and `base32`-encoding the result.

(4) means that if we botch the canonical YAML, the hashes can change.  The anchors could cause this, so I'll at least hard-disable them as described in the link above.

**Comment by alalazo** [comment]
@tgamblin What about the fact that e.g. `py-nose` is listed as a `('build', 'link')` dependency in one of the spec above, while it is only `('build',)` in `py-numpy/package.py`?

**Comment by citibeth** [comment]
Glad that it looks like a solution is at hand!

BTW... would it be more robust if we just hashed the DAG directly, rather
than hashing its YAML representation?

**Comment by davydden** [comment]
> For some reason the other two installations differ in the fact that either py-setuptools or py-nose are considered to be of type=('build', 'link') instead of just type=('build,). This is strange to me, as I can't think of a way in which the edge type can be changed without changing py-numpy/package.py.

This looks similar to https://github.com/LLNL/spack/issues/1457 which has to do with merging of dependency types.

@adamjstewart 

> How do I create and use testing (mock?) packages?

look at this PR for an example https://github.com/LLNL/spack/pull/2064 But I agree that it should be documented at some point.

**Comment by tgamblin** [comment]
@alalazo: I'm not sure about the nose part. But I've got a PR to debug this and at least a fix for the handles.
